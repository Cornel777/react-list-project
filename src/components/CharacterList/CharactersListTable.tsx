import { TableContainer, Paper, Table, TableHead, TableBody } from "@mui/material";
import { FunctionComponent } from "react";
import { Character } from "../../types/Character";
import Homeworld from "../CharacterDetails/components/Homeworld/Homeworld";
import { StyledTableRow, StyledTableCell } from "./styled/styledListTable";
import getIdFromUrl from "./utils/getIdFromUrl";

interface CharactersListTableProps {
  characters: Character[];
  onCharacterNavigation: (id: string) => void;
  onKeyDownNavigation: (e: React.KeyboardEvent, id: string) => void;
}

const CharactersListTable: FunctionComponent<CharactersListTableProps> = ({ 
  characters,
  onCharacterNavigation,
  onKeyDownNavigation
}) => {

  return (
    <TableContainer component={Paper} style={{ maxWidth: '500px' }}>
          <Table size="medium" aria-label="Characters list table">
            <TableHead>
              <StyledTableRow>
                <StyledTableCell>No.</StyledTableCell>
                <StyledTableCell>Name</StyledTableCell>
                <StyledTableCell>Gender</StyledTableCell>
                <StyledTableCell>Home Planet</StyledTableCell>
              </StyledTableRow>
            </TableHead>
            <TableBody>
              {characters.map((character) => {
                const characterId = getIdFromUrl(character.url)
                  return (
                      <StyledTableRow 
                        key={characterId} 
                        tabIndex={0} 
                        onClick={() => onCharacterNavigation(characterId!)}
                        onKeyDown={(e) => onKeyDownNavigation(e, characterId!)}
                      >
                        <StyledTableCell>{characterId}</StyledTableCell>
                        <StyledTableCell>{character.name}</StyledTableCell>
                        <StyledTableCell>{character.gender}</StyledTableCell>
                        <StyledTableCell><Homeworld homeworldUrl={character.homeworld}/></StyledTableCell>
                      </StyledTableRow>
                  )
              })}
           </TableBody>
          </Table>
       </TableContainer>
  )
}

export default CharactersListTable;