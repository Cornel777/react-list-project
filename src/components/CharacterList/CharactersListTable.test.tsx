import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import CharactersListTable from './CharactersListTable';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';


const mockCharacters = [
  {
    name: 'Luke Skywalker',
    url: 'https://swapi.dev/api/people/1/',
    height: '172',
    gender: 'male' as const,
    homeworld: 'https://swapi.dev/api/planets/1/',
    films: [],
    species: [],
    vehicles: [],
    starships: [],
    created: '',
    edited: '',
    birth_year: '',
    eye_color: '',
    hair_color: '',
    skin_color: '',
    mass: ''
  }
];

describe('CharactersListTable', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false
      }
    }
  });

  function renderWithClient(ui: React.ReactElement) {
    const testQueryClient = queryClient;
    const { rerender, ...result } = render(
        <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
    )
    return {
        ...result,
        rerender: (rerenderUi: React.ReactElement) =>
            rerender(
                <QueryClientProvider client={testQueryClient}>{rerenderUi}</QueryClientProvider>
            ),
    }
  }
  
  const defaultProps = {
    characters: mockCharacters,
    onCharacterNavigation: vi.fn(),
    onKeyDownNavigation: vi.fn()
  };

  it('matches snapshot', () => {
    const { container } = renderWithClient(<CharactersListTable {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('renders table with character data', () => {
    renderWithClient(<CharactersListTable {...defaultProps} />);
    expect(screen.getByText('Luke Skywalker')).toBeInTheDocument();
    expect(screen.getByText('male')).toBeInTheDocument();
  });

  it('calls navigation callback on row click', () => {
    renderWithClient(<CharactersListTable {...defaultProps} />);
    const row = screen.getByText('Luke Skywalker').closest('tr');
    fireEvent.click(row!);
    expect(defaultProps.onCharacterNavigation).toHaveBeenCalledWith('1');
  });

  it('calls keyboard navigation on Enter press', () => {
    renderWithClient(<CharactersListTable {...defaultProps} />);
    const row = screen.getByText('Luke Skywalker').closest('tr');
    fireEvent.keyDown(row!, { key: 'Enter' });
    expect(defaultProps.onKeyDownNavigation).toHaveBeenCalledWith(
      expect.any(Object),
      '1'
    );
  });

  it('calls keyboard navigation on Space press', () => {
    renderWithClient(<CharactersListTable {...defaultProps} />);
    const row = screen.getByText('Luke Skywalker').closest('tr');
    fireEvent.keyDown(row!, { key: ' ' });
    expect(defaultProps.onKeyDownNavigation).toHaveBeenCalledWith(
      expect.any(Object),
      '1'
    );
  });

  it('renders table with correct headers', () => {
    renderWithClient(<CharactersListTable {...defaultProps} />);
    expect(screen.getByText('Name')).toBeInTheDocument();
    expect(screen.getByText('Gender')).toBeInTheDocument();
    expect(screen.getByText('Home Planet')).toBeInTheDocument();
  });

  it('maintains accessibility attributes', () => {
    renderWithClient(<CharactersListTable {...defaultProps} />);
    expect(screen.getByRole('table')).toHaveAttribute(
      'aria-label',
      'Characters list table'
    );
    expect(screen.getByText('Luke Skywalker').closest('tr')).toHaveAttribute(
      'tabIndex',
      '0'
    );
  });
});