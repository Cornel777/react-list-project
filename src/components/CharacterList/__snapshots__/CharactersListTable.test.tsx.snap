// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharactersListTable > matches snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); max-width: 500px;"
  >
    <table
      aria-label="Characters list table"
      class="MuiTable-root css-1xwxv7r-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-r0kehe-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
            scope="col"
          >
            No.
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
            scope="col"
          >
            Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
            scope="col"
          >
            Gender
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
            scope="col"
          >
            Home Planet
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-r0kehe-MuiTableRow-root"
          tabindex="0"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            1
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            Luke Skywalker
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            male
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            <div>
              Loading...
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
